### ISSUE TITLE
Indexed cannot be used as integration variable

### ISSUE THREAD

**Issue by Upabjojr** [sympy/sympy#9695]
`Indexed` has the potential for many applications in physics as an integration variable.

```
In [1]: from sympy.tensor import *

In [2]: i = symbols('i', cls=Idx)

In [3]: r = Indexed('r', i)

In [4]: r
Out[4]: r[i]

In [5]: Integral(r, r)
---------------------------------------------------------------------------
ValueError: Invalid limits given: (r[i],)
```

How should it work:
```
>>> integrate(r[i], r[i])
1/2*r[i]**2

>>> integrate(r[i], r[k])
r[i]*r[k]

>>> integrate(x, r[i])
x*r[i]

>>> integrate(sin(r[i]), r[i])
-cos(r[i])
```

**Comment by Upabjojr** [comment]
This is loosely related to:
https://github.com/sympy/sympy/issues/9635

**Comment by Upabjojr** [comment]
@moorepants

**Comment by moorepants** [comment]
Agreed, this is certainly my desired behavior.

**Comment by bjodah** [comment]
I agree that this would be desirable.
One minor question regarding the proposed behavior though:
```
>>> integrate(r[i], r[k])
r[i]*r[k]
```
shouldn't this result require an assumption that `i != k`, so the default here might be Piecewise?

**Comment by Upabjojr** [comment]
> shouldn't this result require an assumption that i != k, so the default here might be Piecewise?

Yes, it should. It will probably make things more complicated to code.

**Comment by Upabjojr** [comment]
I think the way to proceed is as follows:

* allow `Indexed` as a valid symbol in `expr_with_limits`
* create a new integration function, say _integrate_indexed_, specifically dedicated to the integration over indexed symbols.
* this function performs a substitution of all `Indexed` object occurrences in the expression to be integrated, performs the integrals and converts back the values.
* In case of `Indexed` objects with the same `IndexedBase` (e.g. _A[i], A[j]_), perform all possible contractions (in general, if there are _k_ _Indexed_ objects with the same _IndexedBase_, they will be _2^k_), perform an integration for each of them, then create a _Piecewise_ function as a result.

I would put most of the code in the tensor module, in order not to clutter the integration module too much.

**Comment by bjodah** [comment]
I have been giving this some more thought and I think I've changed my mind about the idea of generating Piecewise(...) objects for all the different permutations. We could just ask the user to substitute variables known to be equal to one of the variables, e.g.:
```
(A[i] + A[j]).subs({i: j})  # i and j are known to be equal
```

That is, the implementation assumes all `Idx` instances to be unique.

This would have 2 advantages:
- Implementation will be much easier
- Integrating large expressions with many `Idx` instances will be much faster.

The downside I can think of:
- It might not be immediately apparent for a user that he/she must substitute their expression to have as few unique `Idx` instances as possible. But if we document it well I think it is OK.

This trade off is not unique to `Idx` in any way, the same issue arises when running `integrate` on any expression. (currently `integrate` is a bit flaky in this respect in my experience).

**Comment by Upabjojr** [comment]
> I have been giving this some more thought and I think I've changed my mind about the idea of generating Piecewise(...) objects for all the different permutations.

I don't think they are permutations, I believe they are partitions.

There appear to be some implementations in SymPy:
http://docs.sympy.org/latest/modules/utilities/iterables.html#partitions
(I was pondering about a for loop over _kbins_)

> in general, if there are k Indexed objects with the same IndexedBase, they will be 2^k

I made a mistake in my previous post.

Anyway, I suggest to create a separate temporary integration function, let's say _\_integrate\_indexed_ (not intended for end-users). This would allow to keep experimenting with integrals over indexed symbols (i.e. keep editing it without deprecation phases), and at the same time avoid messing up SymPy's integrator.

So, partition experiment or not?

**Comment by bjodah** [comment]
I think it's a good idea, maybe we could even add a boolean keyword argument along the line of `assume_unique`? That would make integration of expressions with very large number of indices tractable.

Yes, I think you are correct. We will have set partitions with different groupings of indices in which they are assumed equal.

**Comment by normalhuman** [comment]
It seems that #9758 implemented this. Should the issue be closed, or am I missing something?

**Comment by Upabjojr** [comment]
That was a toy implementation. The purpose is to overload the real integral function.
